File: src/main/java/com/examportal/controller/UserController.java
package com.examportal.controller;

import java.security.Principal;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.examportal.dto.UserResponseDTO;
import com.examportal.dto.UserUpdateDTO;
import com.examportal.service.UserService;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;

@RestController
@RequestMapping("/api/users")
//@CrossOrigin(origins = "http://localhost:3000")
@CrossOrigin(origins = "*")
@Tag(name = "User Management", description = "User management APIs")
@SecurityRequirement(name = "bearerAuth")
public class UserController {

    @Autowired
    private UserService userService;

    @GetMapping("/profile")
    @PreAuthorize("hasRole('STUDENT') or hasRole('TEACHER')")
    @Operation(summary = "Get current user profile")
    public ResponseEntity<UserResponseDTO> getUserProfile(Principal principal) {
        UserResponseDTO user = userService.getUserProfile(principal.getName());
        return ResponseEntity.ok(user);
    }
    
    @GetMapping
    @PreAuthorize("hasRole('TEACHER')")
    @Operation(summary = "Get all users - TEACHER only")
    public ResponseEntity<List<UserResponseDTO>> getAllUsers() {
        List<UserResponseDTO> users = userService.getAllUsers();
        return ResponseEntity.ok(users);
    }
    
    @GetMapping("/{id}")
    @PreAuthorize("hasRole('TEACHER')")
    @Operation(summary = "Get user by ID - TEACHER only")
    public ResponseEntity<UserResponseDTO> getUserById(@PathVariable Long id) {
        UserResponseDTO user = userService.getUserById(id);
        return ResponseEntity.ok(user);
    }
    
    @PutMapping("/{id}")
    @PreAuthorize("hasRole('TEACHER')")
    @Operation(summary = "Update user - TEACHER only")
    public ResponseEntity<UserResponseDTO> updateUser(@PathVariable Long id, @Valid @RequestBody UserUpdateDTO updateDTO) {
        UserResponseDTO updatedUser = userService.updateUser(id, updateDTO);
        return ResponseEntity.ok(updatedUser);
    }
    
    @DeleteMapping("/{id}")
    @PreAuthorize("hasRole('TEACHER')")
    @Operation(summary = "Delete user - TEACHER only")
    public ResponseEntity<Void> deleteUser(@PathVariable Long id) {
        userService.deleteUser(id);
        return ResponseEntity.noContent().build();
    }
}

File: Dotnet3/ExamPortal/Controllers/UserController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using ExamPortal.DTOs;
using ExamPortal.Services.Interfaces;
using System.Security.Claims;

namespace ExamPortal.Controllers
{
    [ApiController]
    [Route("api/users")]
    [Authorize]
    public class UserController : ControllerBase
    {
        private readonly IUserService _userService;
        
        public UserController(IUserService userService)
        {
            _userService = userService;
        }
        
        /// <summary>
        /// Get current user profile
        /// </summary>
        [HttpGet("profile")]
        [Authorize(Roles = "STUDENT,TEACHER")]
        public async Task<ActionResult<UserResponseDTO>> GetUserProfile()
        {
            var username = User.FindFirst(ClaimTypes.Name)?.Value;
            if (string.IsNullOrEmpty(username))
                return Unauthorized();
                
            var user = await _userService.GetUserProfileAsync(username);
            return Ok(user);
        }
        
        /// <summary>
        /// Get all users - TEACHER only
        /// </summary>
        [HttpGet]
        [Authorize(Roles = "TEACHER")]
        public async Task<ActionResult<List<UserResponseDTO>>> GetAllUsers()
        {
            var users = await _userService.GetAllUsersAsync();
            return Ok(users);
        }
        
        /// <summary>
        /// Get user by ID - TEACHER only
        /// </summary>
        [HttpGet("{id}")]
        [Authorize(Roles = "TEACHER")]
        public async Task<ActionResult<UserResponseDTO>> GetUserById(long id)
        {
            var user = await _userService.GetUserByIdAsync(id);
            return Ok(user);
        }
        
        /// <summary>
        /// Update user - TEACHER only
        /// </summary>
        [HttpPut("{id}")]
        [Authorize(Roles = "TEACHER")]
        public async Task<ActionResult<UserResponseDTO>> UpdateUser(long id, [FromBody] UserUpdateDTO updateDTO)
        {
            var updatedUser = await _userService.UpdateUserAsync(id, updateDTO);
            return Ok(updatedUser);
        }
        
        /// <summary>
        /// Delete user - TEACHER only
        /// </summary>
        [HttpDelete("{id}")]
        [Authorize(Roles = "TEACHER")]
        public async Task<ActionResult> DeleteUser(long id)
        {
            await _userService.DeleteUserAsync(id);
            return NoContent();
        }
    }
}
